### PR TITLE
miscell fixes in README and images, and the embryo of a containers module

### PR DESCRIPTION
**Pull request by parmentelat**
* the example in the README now works fine
* fix broken `images.remove()`
* new `containers` submodule with `list_containers`, `inspect` and `kill` for starters

this builds on top of PR#18 that I closed before it was merged

### CODE REVIEW

**Comment by parmentelat** [comment]
I have a couple questions regarding the doc that I found here <http://docs.podman.io/en/latest/_static/api.html>

* is this the right place to look ?
* if so what's the difference between the `containers` and `containers (compat)` tabs ?
* same for `images` *vs* `images (libpod)` ?

thanks !

**Comment by parmentelat** [comment]
also I don't know how to handle the failed check, details don't help, any clue appreciated

**Comment by rhatdan** [comment]
LGTM
Although could you fix your commit messages to be less then 80 characters, to remove the ugly line wrap.

@jwhonce @baude @TomSweeneyRedHat PTAL
The meta tests looks like something is broken in CI.

**Comment by TomSweeneyRedHat** [comment]
LGTM
but would like a @jwhonce head nod

**Comment by parmentelat** [comment]
I have rebased and cleaned up on top of the other PR
I have a few other things to propose but I'd rather not mix things up too much before I can get one batch through, please instruct if anything else needs to be tweaked

**Comment by rhatdan** [comment]
Tests are failing.

**Comment by parmentelat** [comment]
not quite sure what is wrong with the tests; simply running `pytest` shows 5/5 passing tests, it looks there's something with the coverage thingy that do not interpret properly ?

**Comment by rhatdan** [comment]
Do you see the error if you run
make validate

**Comment by parmentelat** [comment]
```
make validate
# I have no great ideas on how to install/check for git-validation
Validating vs commit ''
GIT_CHECK_EXCLUDE="./vendor:docs/make.bat" git-validation -run DCO,short-subject,dangling-whitespace -range 911121d43ca2707b2ff5aa7a22342aa74f5b2036..HEAD
/bin/sh: git-validation: command not found
make: *** [.gitvalidation] Error 127
```

**Comment by parmentelat** [comment]
I think I get it: 
```
make unittest
coverage run -m unittest discover -s test/unit
......
----------------------------------------------------------------------
Ran 6 tests in 0.001s

OK
coverage report -m --skip-covered --fail-under=50 --omit=./test/*
Name                            Stmts   Miss  Cover   Missing
-------------------------------------------------------------
<snip>
-------------------------------------------------------------
TOTAL                             160     84    48%

1 file skipped due to complete coverage.
make: *** [unittest] Error 2
```

**`--fail-under=50`** and coverage is **`48%`**  
can you please confirm this is the problem ?

File: podman/__init__.py
Comment (manics) [inline podman/__init__.py:4]: Is this right? `__ALL__` contains `system` whereas this imports it under `podman`

**Comment by jwhonce** [comment]
@parmentelat Please rebase on latest

File: podman/__init__.py
Comment (parmentelat) [reply podman/__init__.py:4]: not quite sure what the intention was with defining __ALL__ here in the first place, but first it should be `__all__` and not `__ALL__`, and second this only impacts `import * from podman` which is a dangerous practice anyways

otoh I was expecting for `import podman` to actually import stuff, hence this approach

**Comment by parmentelat** [comment]
hi

I'm really unsure what to do with that PR; what I need is a basic toolset to list and manage containers; that's what I added on top of the code that I started with 6 months ago, and that's what I've been using in production since, and as you can see the needs are really scarce

I've taken a look at master;  there's still not even a `containers/` submodule in there, and trying to rebase suggests I'm going to have to redo the same work over again as the ground classes have changed

so all in all a little bit of context on a roadmap for this project, and why it is still so embryoid could help build motivation :)

**Comment by rhatdan** [comment]
@parmentelat Most people are using docker-py to talk to the podman socket, at this point.  We are looking for volunteers to build out the full interaction with the podman socket for the full "libpod" api.  This is being mainly driven by community at this point, since the core developers have had different priorities.  We are trying to bring more focus on this effort now.  And hope to see it grow going forward.

If you can help to implement some of these missing features, it would be appreciated.

**Comment by parmentelat** [comment]
thanks @rhatdan for the clarification, this is helpful

here you go, rebased and cleaned up

**Comment by rhatdan** [comment]
@mwhahaha PTAL

File: podman/containers/__init__.py
Comment (mwhahaha) [inline podman/containers/__init__.py:69]: api.delete(path, query)

File: podman/images/__init__.py
Comment (mwhahaha) [inline podman/images/__init__.py:21]: Yea this was a typo. I've caught this in #37 and it's fixed.

File: podman/system/__init__.py
Comment (mwhahaha) [inline podman/system/__init__.py:38]: api.get("/info")

**Review by mwhahaha — CHANGES_REQUESTED** [review]

**Comment by parmentelat** [comment]
all right so it sounds like i should wait until #37 gets merged, right ?

**Comment by mwhahaha** [comment]
Yea I think it would be better. I wish this was gerrit instead of github so we could just manage code change dependencies better.

**Comment by jwhonce** [comment]
@parmentelat Sorry, I stepped on some of your containers work in https://github.com/containers/podman-py/pull/38 . I stopped at list_containers() before abusing too much.  I did add the framework for container unit tests.

**Comment by parmentelat** [comment]
well it feels like I'd better wait also for #38 to land then...

**Comment by rhatdan** [comment]
@parmentelat Needs a rebase and #38 has landed.

**Comment by rhatdan** [comment]
@parmentelat Are you still working on this?

**Comment by parmentelat** [comment]
Hi there

ok, so, to be clear: what I had before - I mean 3 days ago when @rhatdan pinged me - was working for me, and I am not actively involved in this area of my system right now

I just took the time to rebase on top of your master branch; the thing is that version is broken for me; this is because of the `all_` argument to `list_containers`

I know `all` is a Python builtin and all, I know about pep008, but please note that it reads
> single_trailing_underscore_: used by convention to avoid conflicts with Python keyword, e.g.
and as you know `all` is just a builtin and not a keyword...

in any case that breaks my code; are you positive you want to go this road ?
I don't know how much you care about being in line with the docker API but their equivalent has no problem with using `all` as a keyword argument
https://docker-py.readthedocs.io/en/stable/containers.html#docker.models.containers.ContainerCollection.list
and it's my feeling that using `all_` instead is going to cause many people to waste their time trying `all` only to realize it's not the right argument name...

so for now I am reverting to using `all` so that my own code keeps on running

besides, I'd like to mention that `containers.list_containers()` is a bit redundant, and that `containers.list()` would do the trick just fine I expect

finally, I'd like to reiterate my comment about defining `__all__`  in the modules; it feels superfluous, what is the objective exactly here ? 

my 2 cents

**Comment by parmentelat** [comment]
all right so, injecting this code into my deployment still breaks with this error

```
Nov 27 09:42:12 bash[3288865]: During handling of the above exception, another exception occurred:
Nov 27 09:42:12 bash[3288865]: Traceback (most recent call last):
Nov 27 09:42:12 bash[3288865]:   File "/usr/local/lib/python3.9/site-packages/nbhosting/courses/model_course.py", line 463, in image_hash
Nov 27 09:42:12 bash[3288865]:     return podman.images.inspect(podman_api, self.image)['Id']
Nov 27 09:42:12 bash[3288865]:   File "/root/podman-py/podman/images/__init__.py", line 22, in inspect
Nov 27 09:42:12 bash[3288865]:     api.raise_image_not_found(e, e.response)
Nov 27 09:42:12 bash[3288865]: AttributeError: 'ApiConnection' object has no attribute 'raise_image_not_found'
```

Looks like this method on ApiConnection has vanished in the meanwhile; is this intentional ?

reverting to my working code for now

**Comment by mwhahaha** [comment]
yes it was renamed to raise_not_found and takes an exception type

**Comment by rhatdan** [comment]
@parmentelat @mwhahaha Is this still being worked on? Or has it be usurped with changes in master?

**Comment by mwhahaha** [comment]
This should be closed as I took these commits and used it to do #44

**Pull request closed without merging**
